azure/services/agentpools/agentpools.go

```diff
@@ -133,10 +134,16 @@ func (s *Service) Reconcile(ctx context.Context) error {
 			},
 		}
 
+		if profile.EnableAutoScaling != nil && *profile.EnableAutoScaling {
+			// unset the nodepool count if autoscaling is enabled
```

Comment:
Is setting the count to nil idiomatic? It would seem more intuitive to simply inherit the current (at time of reconcile) count value from AKS itself. Not sure if there's a precedent for capz to passively consume (rather than enforce) state changes, but setting it to nil seems weird.

Lemme know if I'm missing something!

Reply:
the issue with setting any value is that we potentially end up in a race condition with cluster autoscaler. Where it scaled up or down a node and try to undo it with our updates later. 

I did a test with not setting the count and it didn't remove the count from Azure API, but it just didn't change it.  
```
        "count": 2,
         ...
        "maxPods": 60,
        "type": "VirtualMachineScaleSets",
        "maxCount": 999,
        "minCount": 0,
        "enableAutoScaling": true,
```

I think the intention is that behaviour, that we let CA do the job and don't interfere with the count. 
We do have precedent to passively consume in CAPZ but we introduce a possible race condition knowing there also. I feel here we can avoid it. But if we agree otherwise also sounds good.

Reply:
What if we did this instead earlier in the code flow, immediately after we have determined that this is not a "create" operation?:

```
		if to.Bool(profile.EnableAutoScaling) {
			if agentPoolSpec.Replicas != to.Int32(existingPool.Count) {
				s.scope.SetAgentPoolReplicas(to.Int32(existingPool.Count))
				log.V(2).Info("AKS AutoScaling enabled, updating Replicas to %d for Pool %s", agentPoolSpec.Replicas, agentPoolSpec.Name)
				log.V(2).Info("After ensuring that pool count is synchronized between capz and AKS, we will apply any other property updates in a future reconciliation")
				return nil
			}
		}
```

Does that break too many architectural boundaries?

cc @CecileRobertMichon @devigned

Reply:
even if `agentPoolSpec.Replicas == to.Int32(existingPool.Count)` doesn't guarantee that between the fetch of the existing pool and the update call to Azure API  cluster autoscaler might have changed the count. If we later set the count in the update request.

But I like `s.scope.SetAgentPoolReplicas(to.Int32(existingPool.Count))` which we should def do also 👍

Reply:
stepping back, another approach that might be cleaner is not to allow setting replicas as it is optional, when autoscaling is enabled. 
we can add that validation for autoscaling in our webhooks.

Reply:
as updating replicas count to spec feels strange. We already have the actual count of nodes in status through cluster API which uses the clusterClient to fetch the actual nodes.

Reply:
Looking at the CAPI + autoscaler integration, cluster-autoscaler takes control of spec.Replicas in that case. I think the issue here is that the AKS autoscaler is directly interacting with the Azure resources, not going through CAPZ CRDs to scale up/down.

I think this discussion is very similar to https://github.com/kubernetes-sigs/cluster-api/pull/5685 where there is a discussion ongoing about pro vs. against setting replicas to nil when autoscaling through the cloud provider is engaged. 

If we do decide to go the nil replicas route, I do agree it would be a better experience to not allow setting replicas at all, rather than overriding the replica count to nil and potentially silently ignoring manual replica count updates.

Reply:
We need to use the `Replicas` field during cluster or node pood creation even if autoscaling is enabled, right?

Can we simply allow `Replicas` to be a configurable field during create only, and then mark it as read-only thereafter? Whether or not to update it during reconciliations I suppose is up to whether or not we want to keep the field "current" with reality so that it can be used as a simple read-only node count property.

Reply:
> We need to use the Replicas field during cluster or node pood creation even if autoscaling is enabled, right?

we can skip it as it's optional, the default value on Azure API is 1. So we can delegat that to Azure API. Even if the nodepool starts from 0 nodes, with autoscaling enabled it should not be a problem. 

> Can we simply allow Replicas to be a configurable field during create only, and then mark it as read-only thereafter?

that is almost what this PR intention was. I didn't change the logic for `create`  but only when doing `update`, ignore(set to nil) the replicas count if autoscaling is enabled.

> If we do decide to go the nil replicas route, I do agree it would be a better experience to not allow setting replicas at all, rather than overriding the replica count to nil and potentially silently ignoring manual replica count updates.

This is a certain use-case when node autoscaling is enabled, so I feel going with this option would be the best. Given all other options.

Reply:
picking this up again.
not allowing setting replicas would be in cluster API level. But we could fail the cluster autoscaling feature validation if the machinepool object has replicas. 

the workflow of updating the replicas with the actual node count of VMSS can be done in separate PR? if we decide to go down that path or even use https://github.com/kubernetes-sigs/cluster-api/pull/5685